BootloaderCommonPkg/Library/PagingLib/PagingMap.c

```diff
@@ -44,60 +50,100 @@ MapMemoryRange (
   )
 {
   UINTN          Idx;
+  UINTN          End;
   UINTN          EntryNum;
-  UINTN         *PageTable;
+  UINTN         *PageTable1G;
+  UINTN         *PageTable2M;
   UINTN         *PageTable4K;
-  UINT32         PageBase;
+  UINTN          PageBase;
   UINT32         Attribute;
-  UINT32         Address;
+  UINTN          Address;
   UINT32         PteOffset;
   UINT32         PdeOffset;
-  UINT32         Alignment;
+  UINT32         PdpOffset;
+  UINTN          Alignment;
   UINT32         PageBits;
 
   Attribute    = IA32_PG_P | IA32_PG_RW | IA32_PG_AC;
 
-  // Define CAR region and memory mapping
-  if (((Ranges[0].Start & 0xFFF) != 0) || ((Ranges[0].Limit & 0xFFF) != 0xFFF)) {
+  // Check for Start and Limit PageSize alignment
+  if (((Ranges[0].Start & (Ranges[0].PageSize - 1)) != 0) || ((Ranges[0].Limit & (Ranges[0].PageSize - 1)) != (Ranges[0].PageSize - 1))) {
     return EFI_INVALID_PARAMETER;
   }
 
-  PageBits  = GetPageShiftBits ();
-  if (PageBits == 21) {
-    // 64 bit mode
-    PteOffset = 6 * EFI_PAGE_SIZE;
-    PdeOffset = 2 * EFI_PAGE_SIZE;
-    EntryNum  = 512;
-  } else {
-    // 32 bit mode
-    PteOffset = 1 * EFI_PAGE_SIZE;
-    PdeOffset = 0;
-    EntryNum  = 1024;
-  }
-
-  Alignment = (UINT32) LShiftU64 (1, PageBits) - 1;
-  PageBase  = Ranges[0].Start & ~Alignment;
-  if (((Ranges[0].Start & ~Alignment) != PageBase) || ((Ranges[0].Limit & ~Alignment) != PageBase)) {
+  // Check for Mapping PageSize alignment
+  if ( (Ranges[0].Mapping & (Ranges[0].PageSize - 1)) != 0) {
     return EFI_INVALID_PARAMETER;
   }
 
-  // Create 4KB pages and remap the CAR region into a different memory location
-  PageTable4K = (UINTN *)((UINTN)PageBuffer + PteOffset);
-  Address = PageBase;
-  for (Idx = 0; Idx < EntryNum; Idx++) {
-    if ((Address >= Ranges[0].Start) && (Address <= Ranges[0].Limit)) {
-      PageTable4K[Idx]   = Ranges[0].Mapping + (Attribute | IA32_PG_PD);
-      Ranges[0].Mapping += SIZE_4KB;
+  switch (Ranges[0].PageSize) {
+  case SIZE_4KB:
+    PageBits  = GetPageShiftBits (PDE_PG_LVL);
+    if (PageBits == 21) {
+      // 64 bit mode
+      PteOffset = 6 * EFI_PAGE_SIZE;
+      PdeOffset = 2 * EFI_PAGE_SIZE;
+      EntryNum  = 512;
     } else {
-      PageTable4K[Idx]   = Address + (Attribute | IA32_PG_PD);
+      // 32 bit mode
+      PteOffset = 1 * EFI_PAGE_SIZE;
+      PdeOffset = 0;
+      EntryNum  = 1024;
     }
-    Address += SIZE_4KB;
+    PageTable4K = (UINTN *)((UINTN)PageBuffer + PteOffset);
+    PageTable2M = (UINTN *)((UINTN)PageBuffer + PdeOffset);
+    break;
+  case SIZE_1GB:

```

Comment:
Where are we using SIZE_1GB?

Reply:
Platform code can call this function with Range[].PageSize = SIZE_1GB, while trying to map > 4GiB PCI resources.

Reply:
Got it. Thanks.